source/common/buffer/buffer_impl.h

```diff
@@ -297,13 +310,35 @@ class Slice {
    * Call all drain trackers associated with the slice, then clear
    * the drain tracker list.
    */
-  void callAndClearDrainTrackers() {
+  void callAndClearDrainTrackersAndCharges() {
     for (const auto& drain_tracker : drain_trackers_) {
       drain_tracker();
     }
     drain_trackers_.clear();
+
+    if (account_) {
+      account_->credit(capacity_);
+    }
+  }
+
+  /**
+   * Charges the provided account for the resources if we're not already
+   * charging for this slice, the account still exists, and the slice owns
+   * backing memory.
+   */
+  void chargeAccountIfValid(const AccountSharedPtr& account) {
```

Comment:
nit: trying to think of a better name for this method, possibly tryChargeAccount or maybeChargeAccount